web_src/js/components/DiffFileTree.vue

```diff
@@ -0,0 +1,129 @@
+        // Split file into directories
+        const splits = file.Name.split('/');
+        let index = 0;
+        let parent = null;
+        let isFile = false;
+        for (const split of splits) {
+          index += 1;
+          // reached the end
+          if (index === splits.length) {
+            isFile = true;
+          }
+          let newParent = {
+            name: split,
+            children: [],
+            isFile
+          };
+
+          if (isFile === true) {
+            newParent.file = file;
+          }
+
+          if (parent) {
+            // check if the folder already exists
+            const existingFolder = parent.children.find(
+              (x) => x.name === split
+            );
+            if (existingFolder) {
+              newParent = existingFolder;
+            } else {
+              parent.children.push(newParent);
+            }
+          } else {
+            const existingFolder = result.find((x) => x.name === split);
+            if (existingFolder) {
+              newParent = existingFolder;
+            } else {
+              result.push(newParent);
+            }
+          }
+          parent = newParent;
+        }
+      }
+      const mergeChildIfOnlyOneDir = (entries) => {
+        for (const entry of entries) {
+          if (entry.children) {
+            mergeChildIfOnlyOneDir(entry.children);
+          }
+          if (entry.children.length === 1 && entry.children[0].isFile === false) {
+            // Merge it to the parent
+            entry.name = `${entry.name}/${entry.children[0].name}`;
+            entry.children = entry.children[0].children;
+          }
+        }
+      };
+      // Merge folders with just a folder as children in order to
+      // reduce the depth of our tree.
+      mergeChildIfOnlyOneDir(result);
+      return result;
+    }
+  },
+
+  mounted() {
+    // ensure correct buttons when we are mounted to the dom
+    this.adjustToggleButton(this.fileTreeIsVisible);
+    document.querySelector('.diff-toggle-file-tree-button').addEventListener('click', this.toggleVisibility);
+  },
+  unmounted() {
+    document.querySelector('.diff-toggle-file-tree-button').removeEventListener('click', this.toggleVisibility);
+  },
+  methods: {
+    toggleVisibility() {
+      this.updateVisibility(!this.fileTreeIsVisible);
+    },
+    updateVisibility(visible) {
+      this.fileTreeIsVisible = visible; // toggle the visibility
+      localStorage.setItem(LOCAL_STORAGE_KEY, this.fileTreeIsVisible);
+      this.adjustToggleButton(this.fileTreeIsVisible);
+    },
+    adjustToggleButton(visible) {
+      const icons = document.querySelectorAll('.diff-toggle-file-tree-button .icon'); // icons[0] is "ToShow", icons[1] is "ToHide"
+      icons[0 + visible].classList.remove('hide'); // 0+true=1 (show the "ToShow"), 0+false=0 (show the "ToHide")
+      icons[1 - visible].classList.add('hide'); //    1-true=0 (hide the "ToHide"), 1-false=1 (hide the "ToShow")
```

Comment:
Here is a trick, I am not sure whether people like it. If it's not liked, it could be rewritten to other styles. 😄